### ISSUE TITLE
A11y Audit Issues

### ISSUE THREAD

**Issue by joshreisner** [code4recovery/tsml-ui#263]
### Discussed in https://github.com/code4recovery/12-step-meeting-list/discussions/952

<div type='discussions-op-text'>

<sup>Originally posted by **Studio88-KM** October 13, 2022</sup>
Hey guys, like I've said so much, great job on this plugin and thank you so much for creating it and making it available to so many! 

I just had a web company who specializes in accessibility do an audit on a website that uses the 12 Meetings plugin and there were many issues. Would you be willing to look at the list and fix? 

Meeting search was confusing:

1. The meeting filters fields are not combined in a field set so it’s not obviously apparent that this search is distinguishable from the full site search in the header.
2. Status message is missing as the results change - nothing is announced to screen readers.
3. Dropdowns are coded as links not radio buttons (which is how they behave) - they are read out as links to the screen reader.
4. Make the number of results more meaningful. I.e., (7) should have an aria-label or hidden screen reader text so that it reads out (7 meetings) or (7 results).
5. When clicking on a meeting to get more information, the meeting loads in the same page and it’s not clear that the information loaded - needs an auditory announcement that the meeting information is loaded or focus to be shifted to the heading on that page.
6. There are elements in the meeting single (under the H3 for the days) that have been set up in lists that are not lists. This is confusing to screen reader users.
7. Maps are not helpful for screen reader users so the toggle to view the meeting map and maps on individual pages would be better hidden from screen readers with aria-hidden=”true”.
8. Allow the option for and H2

</div>

**Comment by joshreisner** [comment]
Thanks @Studio88-KM for the feedback. 

1. We can surround the search in a `fieldset` tag - is that all? Or do they envision changing the whole control row to a be one big form?
2. I'm not clear on how to announce a status message to screen readers as the results change, could you clarify? 
3. The dropdown triggers are coded as `button` elements - do they mean the links inside the dropdown menus should be coded as radio buttons? It doesn't really work that way, they are not part of a form that is submitted.
4. Ok
5. Ok, I can try focusing on the `h1`
6. I'm not sure I agree with that. The meetings listed below Monday are, I think, properly in a list. Could we get some more context around that?
7. Ok!
8. I'm not sure what this means, could you clarify?

**Comment by Studio88-KM** [comment]
Sure! I am sharing this with the Amber, the accessibility company rep so she can clarify. 

#8. Wondering if you can add a possibility for an H2. Right now the page is dynamically generated so can't easily add an H2 that we can see.

**Comment by joshreisner** [comment]
Thanks!

Do you mean insert a new `h2` in the meeting finder? 

Or would this be more of a "start at heading level" setting so the current `h1` title becomes an `h2` and each subsequent heading is demoted a step? (You might want this if you already had an h1 in your site that you didn't want TSML UI competing with.)

**Comment by Studio88-KM** [comment]
Hmmmm.....great question and options. 

I think just inserting an H2 in the meeting finder. For accessibility, there needs to be at least one header other than the H2 on each page. Amber might say something different but I think this is the case.

**Comment by Studio88-KM** [comment]
Woops, I meant other than the H1. I just edited that above.

**Comment by joshreisner** [comment]
I've made some updates:

* there is a `fieldset` now in the search input form
* dropdown badges now have aria-labels, eg `10 results`
* when navigating to the meeting screen it focuses on the h1
* maps are `aria-hidden="true"`
* meeting detail screens have several `h2`s, eg `Meeting Information`
* map info card locations have `h2`s

**Comment by joshreisner** [comment]
I'm not really clear on where an H2 would go on the main meetings page. What would it say? 

Potentially you could add an H2 below the meeting finder using the Meetings Bottom widget area that said something like "Can't find what you're looking for?" with a Contact Us CTA. This would be way down the page but right at the top if they were searching for something and didn't have many results.

**Comment by Studio88-KM** [comment]
That's an idea. I tested it and not quite sure why the Meetings Bottom widget is ending up on the top of the page but it's probably the site. I'll play with it.

**Comment by amberhinds** [comment]
Hi @joshreisner, I'm happy to help clarify needs.

> We can surround the search in a fieldset tag - is that all? Or do they envision changing the whole control row to a be one big form?

Ideally, from a usability standpoint, yes, this should be a form. You can still dynamically submit the content as selections are made if you want, but it would be best for users of assistive technology if there was an option to turn off auto-submit and use a submit button instead on the filters. 

> I'm not clear on how to announce a status message to screen readers as the results change, could you clarify?

You can use aria-live to provide announcements to screen reader users when the content of the page changes. You would want to add these both when the results in the meetings table filter and also when a meeting is clicked and the information page for that meeting opens. Here are two resources that you may find helpful:
- [ARIA live regions](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Live_Regions)
- [Provide notification of dynamic changes to content](https://accessibility.huit.harvard.edu/provide-notification-dynamic-changes-content)

> The dropdown triggers are coded as button elements - do they mean the links inside the dropdown menus should be coded as radio buttons? It doesn't really work that way, they are not part of a form that is submitted.

This references items that are like this and appear in the dropdown menus, yes. Here is an example:
`<div class="dropdown-menu my-1 show" aria-labelledby="time">
    <a class="dropdown-item active bg-secondary text-white" href="https://wordpress-772911-2627160.cloudwaysapps.com/index.php/meetings/">Any Time</a>
    <div class="dropdown-divider"></div>
    <a class="align-items-center d-flex dropdown-item justify-content-between" href="https://wordpress-772911-2627160.cloudwaysapps.com/index.php/meetings/"><span>Morning</span><span aria-label="25 results" class="badge bg-light border ms-3 text-dark">25</span></a>
    <a class="align-items-center d-flex dropdown-item justify-content-between" href="https://wordpress-772911-2627160.cloudwaysapps.com/index.php/meetings/"><span>Midday</span><span aria-label="21 results" class="badge bg-light border ms-3 text-dark">21</span></a>
    <a class="align-items-center d-flex dropdown-item justify-content-between" href="https://wordpress-772911-2627160.cloudwaysapps.com/index.php/meetings/"><span>Evening</span><span aria-label="49 results" class="badge bg-light border ms-3 text-dark">49</span></a>
    <a class="align-items-center d-flex dropdown-item justify-content-between" href="https://wordpress-772911-2627160.cloudwaysapps.com/index.php/meetings/"><span>Night</span><span aria-label="8 results" class="badge bg-light border ms-3 text-dark">8</span></a></div>`

It is incorrect to code these as links as they don't navigate the user anywhere on the page (or to another page), rather they trigger filtering of the meetings in the table. They behave like radio buttons because you can only select one at a time (rather than checkboxes where you would be able to check multiple).

Ideally, these should be recoded into radio buttons so that they have correct HTML semantics. You would wrap them all in a fieldset so that the label is read out and the user also knows how many options there are in the group. If you aren't sure what this should sound like, [listen to the screen reader going through the radio buttons in this video](https://youtu.be/RgVkOmgI0Zs).

If you don't want to completely rebuild it you can communicate to screen readers that these behave like radio buttons with role=radio and can toggle the select state using the aria-checked=true/false. Here is a resource:
- [ARIA radio role](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/radio_role)

> The meetings listed below Monday are, I think, properly in a list. Could we get some more context around that?

The improper use of list is showing up in the meetings grid:
`<div class="meetings d-grid gap-2">
    <div>
        <h3 class="mb-1 mt-2">Tuesday</h3>
        <ol class="list-unstyled">
            <li class="d-flex flex-row gap-2 justify-content-between m-0">
                <div class="text-muted text-nowrap">5:45 PM</div>
                <div class="flex-grow-1">The DA Promises</div>
                <div class="align-items-start d-flex gap-1 justify-content-end pt-1"><small class="align-items-center d-flex flex-row float-end gap-2 px-2 py-1 rounded text-sm in-person"><svg data-testid="icon-geo" fill="currentColor" height="13" viewBox="0 0 16 16" width="13">[...]</path></svg></small></div>
            </li>
        </ol>
    </div>
    <div>
    <h3 class="mb-1 mt-2">Wednesday</h3><ol class="list-unstyled"><li class="d-flex flex-row gap-2 justify-content-between m-0"><div class="text-muted text-nowrap">6:00 PM</div><div class="flex-grow-1"><a href="https://wordpress-772911-2627160.cloudwaysapps.com/index.php/meetings/?meeting=business-debtors-anonymous-bda">Business Debtors Anonymous - BDA</a></div><div class="align-items-start d-flex gap-1 justify-content-end pt-1"><small class="align-items-center d-flex flex-row float-end gap-2 px-2 py-1 rounded text-sm in-person"><svg data-testid="icon-geo" fill="currentColor" height="13" viewBox="0 0 16 16" width="13"><path fill-rule="evenodd" d="M8 1a3 3 0 1 0 0 6 3 3 0 0 0 0-6zM4 4a4 4 0 1 1 4.5 3.969V13.5a.5.5 0 0 1-1 0V7.97A4 4 0 0 1 4 3.999zm2.493 8.574a.5.5 0 0 1-.411.575c-.712.118-1.28.295-1.655.493a1.319 1.319 0 0 0-.37.265.301.301 0 0 0-.057.09V14l.002.008a.147.147 0 0 0 .016.033.617.617 0 0 0 .145.15c.165.13.435.27.813.395.751.25 1.82.414 3.024.414s2.273-.163 3.024-.414c.378-.126.648-.265.813-.395a.619.619 0 0 0 .146-.15.148.148 0 0 0 .015-.033L12 14v-.004a.301.301 0 0 0-.057-.09 1.318 1.318 0 0 0-.37-.264c-.376-.198-.943-.375-1.655-.493a.5.5 0 1 1 .164-.986c.77.127 1.452.328 1.957.594C12.5 13 13 13.4 13 14c0 .426-.26.752-.544.977-.29.228-.68.413-1.116.558-.878.293-2.059.465-3.34.465-1.281 0-2.462-.172-3.34-.465-.436-.145-.826-.33-1.116-.558C3.26 14.752 3 14.426 3 14c0-.599.5-1 .961-1.243.505-.266 1.187-.467 1.957-.594a.5.5 0 0 1 .575.411z"></path></svg></small></div></li></ol></div></div>`

**Comment by amberhinds** [comment]
> on turning off auto-submit, should this potentially also apply to the radio buttons, if it were all one big form?

Yes. It can be very disorienting for blind or low vision users if the content changes without hitting a submit button, so it would ideally give them the ability select options and then submit to filter.

This could be a submit button for the entire form or it could be separate submits for each dropdown.  (Here's an example of that: https://mpvre.com/available-properties/.)

>on the buttons, it's also a little tricky because they can, in a not-obvious feature, behave like checkboxes if the command key is down.

Yes, we didn't even realize that was possible. If these were coded as actual checkboxes or even if you made the links appear to have checkboxes to the left of them, that would be more obvious. The example I shared above is one that had divs styled to look like checkboxes (not great HTML, but coded many years ago) and was remediated to add all the aria-roles that would  treat them like check boxes.  I think it's a decent comparison to what you might need to do if you inspect the code.

**Comment by Studio88-KM** [comment]
Hey @joshreisner , checking in to see what your thoughts are about these suggestions above and if you think you'll be updating the plugin.

**Comment by joshreisner** [comment]
Hi @Studio88-KM, yes, still working on it, no specific ETA for changes yet.

**Comment by Studio88-KM** [comment]
Got it. Thanks! I'll let them know. Appreciate it.

**Comment by joshreisner** [comment]
Ok! In addition to the changes made last month:

- there is a fieldset now in the search input form
- dropdown badges now have aria-labels, eg 10 results
- when navigating to the meeting screen it focuses on the h1
- maps are aria-hidden="true"
- meeting detail screens have several h2s, eg Meeting Information
- map info card locations have h2s

Also now rolling out:

- `aria-live` on the h1 title
- updating dropdown links to use `button`

These dropdown buttons are not ready to be a part of the form yet, that will have to be a part of a larger refactor for #171 . The issue is that currently they can be either checks or radios. I think the solution will be to make them always checks, and dispense with the current `metaKey` functionality, but we'll need to have a UI change as part of that to make it clear to the user how the dropdowns are now operating.

**Comment by Studio88-KM** [comment]
Gotcha. This is great, thanks Josh!

Is the best way for get updates at this point to subscribe to #171?

**Comment by joshreisner** [comment]
yes 👍
